File: 0684-redundant-connection/0684-redundant-connection.java
class Solution {
  public int[] findRedundantConnection(int[][] edges) {
    int[] sets = new int[edges.length + 1];
    
    for(int[] edge : edges) {
        int u = find(sets, edge[0]);
        int v = find(sets, edge[1]);  
        if(u == v) 
            return edge;
        sets[u] = v;
    }
    
    return new int[]{};
}

int find(int[] sets, int v) {
    return sets[v] == 0 ? v : find(sets, sets[v]);
}
}

File: 0684-redundant-connection/0684-redundant-connection.cpp
class Solution {
public:
   int getParent(vector<int>& parents, int v) {
        if(parents[v]==0) return v;
        return parents[v] = getParent(parents, parents[v]);
    }
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        vector<int> parents(edges.size()+1);
        for(auto& e:edges) {
            int gr1=getParent(parents, e[0]);
            int gr2=getParent(parents, e[1]);
            if(gr1==gr2) return {e[0], e[1]};
            parents[gr1]=gr2;
        }
        return {};
    }
};